### ISSUE TITLE
The new image can't be recognized

### ISSUE THREAD

**Issue by namejm** [afriess/dexif#1]
I compiled the DExView.dproj in Delphi 10.2
Click the Write Small button in DExView.exe, create a new image
but the new image can't be recognized by the image brower
Is the WriteEXIFJpeg function wrong?

**Comment by afriess** [comment]
I have no Delphi (10.2) for testing, so it will be consume some time to veryfy this.

**Comment by afriess** [comment]
Delphi 10.1 Berlin and the Picture from pictures folder,

maybe there a problem with the reading of the size. The goal of this project here is the support for Lazarus/Freepascal. Alle Unittesting is actually done only for Lazarus. The Samples here are from the original dExif Package and (actual) only controlled for compiling.

-- EXIF-Data -------------- 
ErrStr = <none>
Motorola Byte Order
Make = SAMSUNG
Model = SM-G850F
Exposure time = 1/3376 sec
Exposure time = 1/3376 sec
FNumber = F2,2
ISOSpeed Ratings = 40, 0
Date Time Original = 2017-03-15 10:35:11
Focal length = 4,09 mm
Exif Offset = 87
No Thumbnail
 
 -- EXIF Summary -(short)--- 
pexels-photo-355241max.jpeg 2947kB 2017-03-15 10:35:11 0w 0h 
 

 -- EXIF Summary -(long)---- 
File name:      pexels-photo-355241max.jpeg
File size:      2947kB
File date:      2017-08-30 15:59:06
Photo date:     2017-03-15 10:35:11
Make (model):   SAMSUNG (SM-G850F)
Dimensions:     0 x 0
Exposure time:  1/3376 sec
Focal length:   4,09 mm
F number        F2,2
ISO:            40, 0
Flash fired:    No

**Comment by wp-xyz** [comment]
Please try the current revision. There were a couple of issues with Big-Endian Exifs. Also, multivalued tags were not written correctly. If this does not resolve the issue, please let me know where to find the image you tested (it looks familiar, though...).

**Comment by namejm** [comment]
The same problem still exists in Delphi 10.2, but it's OK in Lazrus 1.6.4
I test the image in the samples\pictures folder.
my OS is Win7 64bit

**Comment by wp-xyz** [comment]
> I test the image in the samples\pictures folder.
Ah this one, the name was so familiar...

I still cannot reproduce the issue. I only have Delphi 10.1 Berlin Starter Edition, but since the program runs even with Delphi 7 I would not expect a sudden change when going from 10.1 to 10.2.

You have Windows 64 bit (like me). Are compiling a 32-bit application (like me) or 64-bit?

You are saying, that is image is "not reckognized by the image browser". Which image browser? Give step-by-step instructions for me to reproduce the error.

Can you post the faulty image for me to have a look with the hex editor?

**Comment by namejm** [comment]
I'm sorry, My English is poor, misunderstand your mean.
I compling a 32-bit application on 64-bit OS.
I click One File button to open a image, then click Write Small button to create a new one, when I open it by ACDSee Pro 8, or FastStone Image Viewer, nothing found.
I'm a newcomer in github, can't find your Email, can you give me the one?

**Comment by namejm** [comment]
I'm sorry I have to go to bed because it's late at the night in my time zone, I'll probably reply mail about 8 hours later.

**Comment by wp-xyz** [comment]
![grafik](https://user-images.githubusercontent.com/30792460/29884345-7d3a4b6e-8db3-11e7-9b55-2848b9c8acc0.png)

I just ran the portable version of FastStone, and it is displaying the image correctly, see attachment.

I'm a newbie to githuh, too. But if you can see the screenshot in this post then it works to just paste or drag the image into the edit aread here on github - i.e., where you are typing a message.

**Comment by afriess** [comment]
I think, this is one of the issue: Dimensions: 0 x 0
I look for a solution to setup DUnit in Delphi too, So the SAME tests can be made in Lazarus and Delphi.

Another idea, was the directory refreshed in FastStone, becaus it is possible the programm didnt rescan the directory after creating. To check this, pls close the program and reopen it.

**Comment by namejm** [comment]
the image below is created by DExView.exe whichi is compiled in Delphi 10.2
mybe I should compile exe in Delphi 10.1
![pexels-photo-355241max _smaller](https://user-images.githubusercontent.com/31474998/29902352-3dfce6f0-8e30-11e7-963a-455359c16b26.jpg)

**Comment by namejm** [comment]
I reopen the image brower, but still can't recognize the new image.

**Comment by wp-xyz** [comment]
Yes, I confirm that the image is defective, the header of the APP1 segment is not right - I wonder how this can happen by a change in compiler. Just to make sure: Does the same effect happen if you compile DExView with Delphi 10.1?

**Comment by namejm** [comment]
I'm downing the Delphi 10.1, it will take an hour to finish.

**Comment by wp-xyz** [comment]
Installing 10.2 Starter...

**Comment by wp-xyz** [comment]
Done. I compiled DExView (no problems) and ran it with the "pexels" image. The "smaller" version stored with the "Write small" button is displayed correctly in Windows Explorer and FastStone Image Viewer (in contrast to the one you posted above). So, I don't know what you are doing wrong... Are you sure that you use the code of the most-up-to-date version of THIS repository? There are a lot of changes at the moment, and I know that a previous version had trouble with images having big-endian EXIF (like this one). Other versions of dExif that you may find somewhere else are not compatible with the current work.

**Comment by namejm** [comment]
I use the code of the newest version of this repository, and compied with Delphi 10.1 and 10.2, but still get the image can't be recognized by the image brower.
Below is the compling info, My OS‘s Language is Chinese, System locale language settings is Simple Chinese, Is that relevant?
[CompileErrWith D10.1.txt](https://github.com/afriess/dexif/files/1266578/CompileErrWith.D10.1.txt)

**Comment by wp-xyz** [comment]
Probably. Since I cannot reproduce this issue on my system I'll have to help me with debugging... 

At first, please, update your repository with the version that I committed a minute ago. It contains modified writing code, and we must make sure to use the same code all the time.

Compile and run DExView again. Now load IMG_9438.jpg from folder tests/testpictures/original - it is a little-endian exif picture. Save it with button "Write same" (this does not change the size, it just rewrites the exif). Can you open it? If not, post the created file. Repeat with the "pexels" image that you used before (it has a big-endian exif).

Run our tests, application TestGUIdExif in folder tests (click on one of the green arrows, it does not matter which one). Which tests are failing (purple icon)? (Here: none, i.e. all icons green).

**Comment by namejm** [comment]
I'm sorry the images still can't be recognized(exe compiled with Delphi 10.2), but the TestGUIdExif's result if OK.
here are the images saved with button "Write same"
![img_9438_same](https://user-images.githubusercontent.com/31474998/29919628-ad37e37a-8e7c-11e7-9540-76c239e4637a.jpg)
![pexels-photo-355241max_same](https://user-images.githubusercontent.com/31474998/29919678-d9f3a728-8e7c-11e7-885b-28732abf2023.jpg)

here is the test result:
![20170831184109](https://user-images.githubusercontent.com/31474998/29919758-3571be6e-8e7d-11e7-8bac-c4dae9b1a441.jpg)

**Comment by wp-xyz** [comment]
Oh - I'm sorry. These tests are Lazarus-only. Please run samples/91_simpletest/Delphi/ReadWriteTest, but I fear all tests will fail in your Delphi.

**Comment by wp-xyz** [comment]
When I change the segment marker byte of the EXIF segment in the two files that you submitted the files are readable! We're lucky: this segment is written at a unique location, and there are suspicious intructions: the header to be written is declared as a string constant (#$FF#$E1) - I just think that Delphi extends this to a wide string. 

Please try the new version where this part is coded without using strings. Repeat the "Write same" test procedure.

**Comment by namejm** [comment]
This version is OK, I can see the new image now. Good!

**Comment by wp-xyz** [comment]
Great - never stop learning... Could you still run the ReadWriteTest? There might be similar places in the code...

**Comment by wp-xyz** [comment]
I forgot: ReadWriteTest is a special test program for Lazarus AND Delphi, it is a bit different from the other test program. Run it and press "Test1", then "Test2" (two different files with test parameters). In each case, the tests are displayed in the list below.  A passed test is marked by a green checkmark, a failed test by a  red x.

**Comment by namejm** [comment]
ReadWriteTest dead :(
![20170831202540](https://user-images.githubusercontent.com/31474998/29923081-c5335c1c-8e8a-11e7-9809-775a94238081.jpg)

ReadWriteTest.exe Compiled with Lazarus:
![20170831210706](https://user-images.githubusercontent.com/31474998/29924797-c3d898c2-8e90-11e7-9638-50da77ce3232.jpg)
![20170831210926](https://user-images.githubusercontent.com/31474998/29924783-b3c03030-8e90-11e7-82de-e26e0a1d7801.jpg)

**Comment by namejm** [comment]
Find a problem: when I use MakeCommentSegment to add a comment section, I can't see the Comment in the new image.

the code likes this:

procedure TfrmMain.lsv_ImagesClick(Sender: TObject);
var
  curItem: TListItem;
  ImgPath: string;
  ImgData: TImgData;
begin
  curItem := nil;
  try
    ScreenToClient(Mouse.CursorPos).Y);
    curItem := lsv_Images.ItemFocused;
  except
    Exit;
  end;

  ImgData := TImgData.Create;
  ImgData.TraceLevel := 1;  // set detail level for trace string

  if (curItem <> nil) then
  begin
    ImgPath := cmb_FolderPath.Text + '\' + curItem.Caption;
    if ImgData.ProcessFile(ImgPath) then
    begin
      if ImgData.HasComment then    // there was a Comment Section
        ShowMessage(ImgData.Comment)
      else
      begin
        ImgData.MakeCommentSegment('test');   // add a comment section
        ImgData.WriteEXIFJpegTo(ImgPath + '.jpg');
      end;
    end;
  end;

  ImgData.Free;
end;

**Comment by wp-xyz** [comment]
Thank you. As expected...

I don't know how familiar you are with the debugger. So the next instructions are maybe a bit too lengthy, I just want to make sure.

Load the test program TestWriteRead again into Delphi (10.1 or 10.2). Open unit rwMain and go to procedure TMainForm.ExecTest. Set a breakpoint on line "listitem := ListView.Items.Add" (at about line 234) (click on the left gutter bar area of the line to set a breakpoint, a red circle will appear). Now press Ctrl+Alt+w to show the list of watched expressions at the left of the text editor. Right-click on this list, select "Add expression" (or maybe "Add watch" - my Delphi is in German), in the dialog box, type "i" (without quotes) in the line "expression" - this watches the loop variable.

Now run the program. Click "Test1".
The program stops at the breakpoint. Look at the watch window and note the value of i. It should be 3 (because the test parameter file contains some comments and empty rows which are skipped). Then press F9 to go to the next break point which, because of the loop, is in the same line, but with the the loop counter i incremented by 1. Repeat until the program hangs. Tell me the value of i.

If it does not hang abort the program and remove the break point (click on the red circle in the gutter). Now set the breakpoint on line "ImgData.ProcessFile(OutFile)" (about line #253). Run again. If the program hangs now before the breakpoint is reached the problem must be in the previous line "ImgData.WriteEXIFJpegto(Outfile)".

If it does not abort the program again, remove the breakpoint and set it 1 line deeper at line  "j := 0". Repeat the test. If the program hangs before reaching the breakpoint the error must be in ImgData.ProcessFile.

If it does not put the breakpoint on line "testdata := Split(testcases[i]". Run. Press F9 as often as needed until the program hangs. Tell me the value of the loop counter.

If it does not put the breakpoint on the final line "ExifTabControlChange(nil)" (remove the old breakpoint). When you run the program it should reach this line. Press F7 (not F9!) to go into this procedure. Then press F8 (not F7, not F9) to advance line by line. Tell me at which the programs hangs.

Sorry for these long instructions, but without debugger it is very unlikely to find the bug.

**Comment by namejm** [comment]
After set breakpoint on line 234, I found the value of i changed from 3 to 27, then 30,31, then exe hung

**Comment by wp-xyz** [comment]
Good. This means that the Chinese Delphi cannot read the ansi parameter files. I modified the reading code a bit such that it is still compatible with Delphi7. Could you try the current revision and report back if you can get past line 234 now.

**Comment by namejm** [comment]
"listitem := ListView.Items.Add" in this version is not 234 but 249, I set breakpoint on 249, after i=31, the exe dead.

**Comment by wp-xyz** [comment]
But it displays all i values (except for two which are commented out near the end)?

Then, if you put a breakpoint at 265 ("WriteExifJpegTo") the program should stop there. When you press F8, does it hang or does the debugger get past this line. What about the next one ("Processfile")? Just proceed line by line by pressing F8 and find the line in which the program hangs.

**Comment by namejm** [comment]
when run at 268 ("ImgData.ProcessFile(OutFile);"), the program hangs.

**Comment by wp-xyz** [comment]
Oh, this is the central routine of dEXIF, it covers everything... Let me think about it.

Maybe one more simple test: Open the parameter file "testcases1.txt" in an editor, it is in "samples/91SimpleTest/common" and add a semicolon (;) in front of every line. This means: this line will be skipped. Since all lines are skipped the program will write back the original exif information but using the routines of dExif. Does the program still hang? You must click "Test 1" to use the file "testcases1.txt".

If not, the hang is caused by my way of changing tags within the exif structure. Remove the semicolons one by one (but keep only a single line uncommented this way) and repeat until you find which parameter line causes the hang.

**Comment by namejm** [comment]
When I add semicolon(;) in front of every line in the txt file, the program is OK
the result like this:
![20170901000331](https://user-images.githubusercontent.com/31474998/29933147-35af8cfe-8ea9-11e7-8cf3-0f648f19206c.jpg)

**Comment by wp-xyz** [comment]
Good - not that bad...

I know it is getting late for you, but could you do now the second part of my previous post, i.e remove the semicolon from the first parameter line ("Comment") and repeat. If the program does not hang add the seimicolon again, and remove it from the second line. etc. There must be (at least) one parameter which, when inserted into the exif structure, causes the program to hang.

**Comment by namejm** [comment]
the "Comment" cause hang
the MaxApertureValue/GPSLatitude/GPSLongitude cause fail but not hang
I'm sorry  it's time to go to bed now, I'll reply after about 8 hours later. See you tomorrow.

**Comment by wp-xyz** [comment]
Good night

**Comment by wp-xyz** [comment]
I refactored the comment part of dExif. Please test if the current version still hangs when reading the comment segment.

The MaxAperture, GPSLatitude and GPSLongitude fails are probably due to different decimal separators. I changed everything to use a decimal point only - not sure if it will remain like that, but let's keep it at least until your bugs are fixed.

As for your other issue with the comment, you should know that this repository is changing a lot with dExif. The old comment routines are no longer available, they are replaced by a property "Comment" of TImgData which you just read or write:
`ImgData.Comment := 'This is a comment;
s := ImgData.Comment; `
Note also that in addition to this comment which is in the COMMENT segment of the JPEG, there is another type of comment which resides inside the exif structure and is accessible via the property "ExifComment" of the ImgData.ExifObj. It corresponds to the tag "UserComment" - maybe I rename it back. Note that the ExifComment can contain UNICODE text while the other comment is encoded in the default code page of Windows (I don't know yet, how it behaves in Linux which does not have code pages).

**Comment by namejm** [comment]
The comment segment is OK now, but there are some problems like this:
![20170901213204](https://user-images.githubusercontent.com/31474998/29972019-3901c55c-8f5d-11e7-8f1a-c8bc59f973dc.jpg)
![20170901212812](https://user-images.githubusercontent.com/31474998/29972022-3b98750e-8f5d-11e7-903f-fd86817e3e18.jpg)

**Comment by wp-xyz** [comment]
I don't understand: In the first screenshot the test is crashing at "Artist", but in the second one it was able to go beyond this test and crashed in IsoSpeedRatings. How did you achieve this? Or are the crashes sporadic ones?

Does the first crash at "Artist" happen also, if "Artist" is the only test case, i.e. all others are commented out with a semicolon?

I wonder why I don't see these issues. Do they happen in Lazarus as well?

**Comment by namejm** [comment]
The errors occurred in the two images only, they happen same in Lazarus:
![20170901222030](https://user-images.githubusercontent.com/31474998/29974164-9c46809c-8f64-11e7-8050-c95aa4eb7892.jpg)
![20170901222134](https://user-images.githubusercontent.com/31474998/29974154-94247e14-8f64-11e7-98dc-b422bd2299bf.jpg)

**Comment by wp-xyz** [comment]
So let's start all over:
- I suppose that the first error happens in the test case "Artist"
- Add semicolons to all test cases in "testparams1.txt", except for "Artist"
- Open unit rwMain, add breakpoints in the lines 235 "ImgData.ProcessFile(EdTestFile.Text)", 260 "currTagValue := ReadTagValue(tagName)", 264 "WriteTagValue(tagName, newTagValue)", 272 "ImgData.WriteExifJPEG", 275 "ImgData.ProcessFile(OutFile)", 280 "currTagvalue := ReadTagValue(tagname)"
- Compile and run. Press "Test1"
- The program should read the first break point (line 235) without problems
- Press F8 to execute this line. Does the program hang? --> report
- If the program does not hang press F9 to proceed to the next break point
- Press F8 to execute this line. Does the program hang?
- etc...

Tell me the first line in which the program hangs.

Then we can address the other case, ISOSpeedRatings, in the same way.

Sorry to keep you busy with this... Just tell me if you want to stop this.

**Comment by wp-xyz** [comment]
Wait a minute.... I'll post a revisied version which could fix the "Artist" error.

**Comment by wp-xyz** [comment]
OK - the current revision could take care of the first issue. At least it fixes a latent problem

**Comment by namejm** [comment]
When run in line 260, it likes this:
![20170901231452](https://user-images.githubusercontent.com/31474998/29976172-b737a1d6-8f6b-11e7-852c-65a7f7002846.jpg)

when press continue, it likes this:
![20170901231943](https://user-images.githubusercontent.com/31474998/29976276-14ff3afe-8f6c-11e7-9baa-c51606988535.jpg)

**Comment by wp-xyz** [comment]
Is this the "Artist" case? And is this the new version?

**Comment by namejm** [comment]
Use the new version, this is the "Artist" case.

**Comment by wp-xyz** [comment]
Then we have to dig deeper...

- Open "rwMain" and set a breakpoint at line 406 "Result := ImgData.ExifObj.Artist"
- Run, press "Test 1"
- When the program stops at the breakpoint press F7
- This brings us into unit dExif, function TImageInfo.GetArtist. We are on line "begin"
- Press F8 to proceed to the next line "Result := GetValueAsString('Artist')
- Press F7 to get into "GetValueAsString"
- Now press F8 to advance line by line
- The program should crash at one of the lines in this procedure.
- If it does not stop and report back
- If it does add watches for the variables used in the line where the crash occurs
- Abort the program and repeat the test straight to the line where the crash will occur with the next F8
- Send me the line number and the values of the variables.

Its best to do this in Delphi which is more userfriendly for debugging.

**Comment by namejm** [comment]
Crash occurs on line 2888 in dExif.pas, like below, the value of tag=**(1636920, 0, 0, 689, 0, '', '', '', '', '', '', 1903476, $D, 64024)** when run at 2886, when 2888, I can't copy the value:
![20170902000618](https://user-images.githubusercontent.com/31474998/29978249-f1fcecb6-8f72-11e7-9625-fd4eaa29a096.jpg)

**Comment by namejm** [comment]
I'm so sorry I have to go to bed now. See you tomorrow.

**Comment by wp-xyz** [comment]
Strange...

- Do you have the Starter version of Delphi? If not continue with Lazarus

- Repeat the previous test until you're at the line "tag = GetTagByName(ATagName)" which would create the error.
- At first make sure that ATagName has the value 'Artist': Move the mouse over "ATagName" and look at the popup window - it should display "Artist". If not stop and report back.
- If this is ok press F7 (not F8) to get into the function GetTagByName.
- Press F8 until you are at line "i := LookupTag(ATagname)".
- Press F7 to get into "LookupTag".
- Press F8 once to get to the first line ("ATagName := Uppercase(...").
- Move the mouse over fiTagCount. The popup should display 68.
- Set a break point at line 1370 "Result := i" and and at line 1373 "end"
- Press F9 to run until the next breakpoint
- At which breakpoint does it stop?
- at 1373 ("end")? Press Ctrl-F7, enter "fiTagArray[0] in line "Expression" of the "Evaluate/Modify" window, make a screenshot, then enter "fiTagArray[8]" and make another screenshot.
- at 1370? Move the mouse over the variable "i". Does the popup display the value 8? If not which one? Press Ctrl-F7 and enter "fiTagArray[i]" in "Expression", make a screenshot.

**Comment by wp-xyz** [comment]
Could it be that the image that you use for the test is damaged somehow? Please upload it that I can have a look.

Then copy image "tests\testpictures\original\img_9438.jpg" to folder "tests\testpictures" and rename it to "DUTPic03.jpeg". Repeat the test.

**Comment by wp-xyz** [comment]
Please use the new version for the tests proposed. I added a check for a case which results in a similar symptom to yours. It could be that the program behaves differently now (i.e. crashes at different locations or does not reach the breakpoints).

**Comment by namejm** [comment]
I tested pre-version code with the image at **..\..\..\tests\testpictures\original\no_exif.jpg**, so Err occured, when I copy image "tests\testpictures\original\img_9438.jpg" to folder "tests\testpictures" and rename it to "DUTPic03.jpeg", the pre-version code is OK, so I think the pre-version codes have a mistake when the image not has exif.
I'll test the new version soon.

**Comment by namejm** [comment]
the new version has the problem **[dcc32 Error] rwMain.pas(197): E2004 Identifier redeclared: 'jpeg'**, so I edit like this:

![20170902095020](https://user-images.githubusercontent.com/31474998/29992012-5c9ae3d8-8fc4-11e7-86bc-0d032b7289b9.jpg)

When test with **..\..\..\tests\testpictures\DUTPic03.jpeg**, it's OK
When test with **..\..\..\tests\testpictures\original\no_exif.jpg**, it raises Err like this:
![20170902094212](https://user-images.githubusercontent.com/31474998/29991990-d92fbce4-8fc3-11e7-830e-93bb326d19f8.jpg)
![20170902094229](https://user-images.githubusercontent.com/31474998/29991991-dda2cc8a-8fc3-11e7-96fd-af170d51b8d9.jpg)

**Comment by afriess** [comment]
This message is correct, because the testpicture (no_exif.jpg) have no EXIF-structure inside. This picture is for detecting the situation. 

This picture is used to detect, if dExif can handle this situation.

**Comment by wp-xyz** [comment]
Looking at your screenshots again I see now that you changed the hard-coded filenames in the filename edit. Therefore, we were comparing the behavior of the program with different test files - I should have seen that in the first place, but i didn't. The past debug session, however, is no wasted time because it helped to find some bugs which I would not have seem otherwise. Of course, the different test files explain a lot of the differences that we saw, but I am still puzzled why I did not see your initial crash on my Delphi.

I added some improvements to the the current version and did a lot of tests. It should behave correctly now in Lazarus and Delphi except for GPS coordinates which are not correctly written to the file.

The "No exif structure detected" error should not happen again (at least in the test program) because now an empty exif data block is created if it is not found in the source file.
